### PR TITLE
Introduction of Pairing Adapters

### PR DESCRIPTION
**Pull request by aszecowka**
**Description**

Changes proposed in this pull request:

- TBD soon
- ...
- ...

**Related issue(s)**

See ##575

### CODE REVIEW

File: chart/compass/charts/pairing-adapter/templates/service.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  name: {{ template "fullname" . }}
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app: {{ .Chart.Name }}
+    release: {{ .Release.Name }}
+spec:
+  type: ClusterIP
+  ports:
+    - port: 80
+      targetPort: 8080
```
Comment (crabtree) [inline chart/compass/charts/pairing-adapter/templates/service.yaml]: This should correspond with the port from the deployment

File: components/director/cmd/director/main.go [hunk 2]
```diff
@@ -135,6 +140,26 @@ func main() {
 	}
 }
 
+func getPairingAdaptersMapping(filePath string) (map[string]string, error) {
+	if filePath == "" {
+		log.Infof("No configuration for pairing adapters")
+		return nil, nil
+	}
+	file, err := os.Open(filePath)
```
Comment (crabtree) [inline components/director/cmd/director/main.go]: In this approach, you need to close the file. Maybe you can use `ioutil.ReadFile` instead.

File: components/director/internal/domain/onetimetoken/service.go [hunk 3]
```diff
@@ -48,6 +84,20 @@ func (s service) GenerateOneTimeToken(ctx context.Context, id string, tokenType
...
 
+	if tokenType == model.ApplicationReference {
+		app, err := s.appSvc.Get(ctx, id)
+
+		if err != nil {
+			return model.OneTimeToken{}, errors.Wrapf(err, "while getting application [id: %s]", id)
+		}
+
+		if app.IntegrationSystemID != nil {
+			if adapterURL, ex := s.intSystemToAdapterMapping[*app.IntegrationSystemID]; ex {
```
Comment (crabtree) [inline components/director/internal/domain/onetimetoken/service.go]: Maybe we should stick to the pattern with `ok`. To be honest `ex` for me points more to the exceptions.
```suggestion
			if adapterURL, ok := s.intSystemToAdapterMapping[*app.IntegrationSystemID]; ok {
```

File: components/director/internal/domain/onetimetoken/service.go [hunk 4]
```diff
@@ -56,6 +106,61 @@ func (s service) GenerateOneTimeToken(ctx context.Context, id string, tokenType
...
+	}
+	graphqlApp := s.appConverter.ToGraphQL(&app)
+	data := pairing.RequestData{
+		Application: *graphqlApp,
+		Tenant:      extTenant,
+	}
+
+	var externalToken string
+	err = retry.Do(func() error {
+		asJSON, err := json.Marshal(data)
```
Comment (crabtree) [inline components/director/internal/domain/onetimetoken/service.go]: The request body preparation can be done outside the retry block.

File: components/director/internal/domain/root_resolver.go [hunk 5]
```diff
@@ -115,11 +116,12 @@ func NewRootResolver(transact persistence.Transactioner, scopeCfgProvider *scope
 	healthCheckSvc := healthcheck.NewService(healthcheckRepo)
 	labelDefSvc := labeldef.NewService(labelDefRepo, labelRepo, uidSvc)
 	systemAuthSvc := systemauth.NewService(systemAuthRepo, uidSvc)
-	tokenSvc := onetimetoken.NewTokenService(connectorGCLI, systemAuthSvc, oneTimeTokenCfg.ConnectorURL)
+	tenantSvc := tenant.NewService(tenantRepo, uidSvc)
+	httpClient := http.DefaultClient
```
Comment (crabtree) [inline components/director/internal/domain/root_resolver.go]: Please provide http client with a request timeout specified.

File: components/pairing-adapter/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,32 @@
+FROM golang:1.13.5-alpine3.10 as builder
```
Comment (crabtree) [inline components/pairing-adapter/Dockerfile]:
```suggestion
FROM golang:1.13.5-alpine3.11 as builder
```

File: components/pairing-adapter/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+COPY . ${BASE_APP_DIR}
+
+#
+# Build app
+#
+
+RUN go build -v -o main ./cmd/main.go
+RUN mkdir /app && mv ./main /app/main && mv ./licenses /app/licenses
+
+FROM alpine:3.10
```
Comment (crabtree) [inline components/pairing-adapter/Dockerfile]:
```suggestion
FROM alpine:3.11.3
```

File: components/pairing-adapter/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+LABEL source = git@github.com:kyma-incubator/compass.git
+WORKDIR /app
+
+#
+# Copy binary
+#
+
+COPY --from=builder /app /app
```
Comment (crabtree) [inline components/pairing-adapter/Dockerfile]: I think you may also need the ca-certifiactes package.

File: components/pairing-adapter/cmd/main.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+	cli := adapter.NewClient(client, conf.Mapping)
+
+	h := adapter.NewHandler(cli)
+
+	http.Handle("/adapter", h)
+	http.HandleFunc("/healthz", func(writer http.ResponseWriter, request *http.Request) {
+		writer.WriteHeader(http.StatusOK)
+	})
+	err = http.ListenAndServe(":8080", nil)
+	if err != nil {
```
Comment (crabtree) [inline components/pairing-adapter/cmd/main.go]: why not just call `exitOnError` func?

File: components/pairing-adapter/internal/adapter/client.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+	"net/http"
+	"text/template"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/pkg/errors"
+)
+
+//go:generate mockery -name=Doer -output=automock -outpkg=automock
+type Doer interface {
```
Comment (crabtree) [inline components/pairing-adapter/internal/adapter/client.go]: Maybe rename it to HTTPDoer to be consistent with other usage.

File: components/pairing-adapter/internal/adapter/client.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+	b, err := ioutil.ReadAll(in)
+	if err != nil {
+		return "", errors.Wrap(err, "while reading response body")
+	}
+	respBody := map[string]interface{}{}
+	if err = json.Unmarshal(b, &respBody); err != nil {
+		return "", errors.Wrap(err, "while unmarshalling response body")
+	}
+
+	logrus.Infof("Got response: %s", string(b))
```
Comment (crabtree) [inline components/pairing-adapter/internal/adapter/client.go]:
```suggestion
	logrus.Infof("Got response: %s\n", string(b))
```

File: docs/compass/03-pairing-adapter.md [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+```config.json: '{"{integration system id}":"{pairing adapter URL}"}'```
+
+for example:
+
+```config.json: '{"0c4bfacf-2237-40f1-ab3d-8e35582e26f2":"http://compass-pairing-adapter/adapter"}'```
+
+Communication between Director, Pairing Adapter and External Token Service is presented in the following diagram:
+
+![](./assets/pairing-adapters.svg)
```
Comment (klaudiagrz) [inline docs/compass/03-pairing-adapter.md]: Please explain the steps under the diagram with the ordered list. What does point no. 6 stand for? Another thing, please add the actor that performs action from step no.1. Correct the capitalization as follows: `ConfigMap`, `one-time token`, `response`. Round the connectors and move connector descriptions and steps above/below the connectors themselves. Last but not least: `Management Plane cluster` vs. `Compass cluster`? So far, we used the second one in the docs. 

When correcting the diagram, please refer to the [guidelines](https://github.com/kyma-project/community/blob/master/guidelines/content-guidelines/02-diagrams.md).

File: components/director/internal/domain/root_resolver.go [hunk 13]
```diff
@@ -143,6 +146,12 @@ func NewRootResolver(transact persistence.Transactioner, scopeCfgProvider *scope
 	}
 }
 
+func getHttpClient() *http.Client {
```
Comment (crabtree) [inline components/director/internal/domain/root_resolver.go]: Now you are modifying the default http client which may be not intended. Maybe you can create here separate instance of http client and use it for that purpose.
```go
return &http.Client{
    Timeout: 3 * time.Second,
}
```

**Pull request merged**